### PR TITLE
Add google top site

### PR DESCRIPTION
**Pull request by nbhasin2**

### CODE REVIEW

File: Client/Frontend/Home/ASFooterView.swift [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (nbhasin2) [inline Client/Frontend/Home/ASFooterView.swift:1]: // refactored from Firefox Home ViewController

File: Client/Frontend/Home/ASHeaderView.swift [hunk 2]
```diff
@@ -0,0 +1,109 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (nbhasin2) [inline Client/Frontend/Home/ASHeaderView.swift:1]: // refactored from Firefox Home ViewController

File: Client/Frontend/Home/ASLibraryCell.swift [hunk 3]
```diff
@@ -0,0 +1,63 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (nbhasin2) [inline Client/Frontend/Home/ASLibraryCell.swift:1]: // refactored from Firefox Home ViewController

File: Client/Frontend/Home/LibraryShortcutView.swift [hunk 4]
```diff
@@ -0,0 +1,53 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (nbhasin2) [inline Client/Frontend/Home/LibraryShortcutView.swift:1]: // refactored from Firefox Home ViewController

File: Client/Frontend/Browser/SearchTelemetry/SearchTelemetry.swift [hunk 5]
```diff
@@ -50,4 +50,15 @@ class SearchTelemetry {
...
         GleanMetrics.Search.inContent["\(provider).organic.none"].add()
     }
+    
+    //MARK: Google Top Site SAP
+    func trackGoogleTopSiteTap() {
+        GleanMetrics.Search.googleTopsitePressed["\(SearchEngine.google).\(code)"].add()
+    }
+    
+    func trackGoogleTopSiteFollowOn() {
+        GleanMetrics.Search.inContent["\(SearchEngine.google).in-content.google-topsite-follow-on.\(code)"].add()
```
Comment (nbhasin2) [inline Client/Frontend/Browser/SearchTelemetry/SearchTelemetry.swift:60]: Waiting on this telemetry to be confirmed and then this PR will be good to be opened for review

Comment (nbhasin2) [reply Client/Frontend/Browser/SearchTelemetry/SearchTelemetry.swift:60]: We are still waiting on this particular telemetry to be confirmed in terms of formatting but overall I am opening PR for review.

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift [hunk 6]
```diff
@@ -660,15 +660,23 @@ extension BrowserViewController: WKNavigationDelegate {
...
+            // Special case google followOn search
+            if (tab.urlType == .googleTopSite || tab.urlType == .googleTopSiteFollowOn) && clientValue == code {
+                tab.urlType = .googleTopSiteFollowOn
+                telemetry.trackGoogleTopSiteFollowOn()
             // Check if previous tab type is search
-            if (tab.urlType == .search || tab.urlType == .followOnSearch) && clientValue == code {
+            } else if (tab.urlType == .search || tab.urlType == .followOnSearch) && clientValue == code {
                 tab.urlType = .followOnSearch
                 telemetry.trackSAPFollowOn()
             } else if provider == .google {
```
Comment (nbhasin2) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift:682]: main logic for follow on and tracking tile search

**Review by boek — APPROVED** [review]
Everything makes sense to me. The only concern I have is `setTabSearchType` behavior is getting pretty complex and might be a good first candidate to tease out of BVC 🤔

**Comment by boek** [comment]
Does this require a data review to land?

**Comment by nbhasin2** [comment]
> `setTabSearchType`

Agreed, we can actually move all this to search telemetry. Let me see what I can do about it. 

You are right @boek we need data review which I have the form ready but waiting on Data Science to approve the two events before I submit the form.

**Comment by nbhasin2** [comment]
# Request for data collection review form 
1. What questions will you answer with this data?
	* How many times users use google's pinned top site 
	* How many follow-on searches are performed by google topsite

2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:
	* Provide information essential for advancing a business objective such as supporting OKRs.
	* Help us in knowing how well are sap's are working 

3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
	* No other way to collect this data.

4. Can current instrumentation answer these questions?
	* No, the google top site was just added and we have no existing telemetry for them.

5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement Description | Data Collection Category | Tracking Bug # |
| --- | ----- | ----- |
When google top site is tapped | 2 | https://jira.mozilla.com/browse/FXIOS-1223
When user performes a follow-on search after visiting google from topsite | 2 | https://jira.mozilla.com/browse/FXIOS-1223

6. How long will this data be collected? Choose one of the following:
	* I want this data to be collected for 6 months initially (potentially renewable).

7. What populations will you measure?
All countries, except for ["CN", "RU", "TR", "KZ", "BY"] and channels that have data collection on.

8. If this data collection is default on, what is the opt-out mechanism for users?
	* Toggle switch for 'Send Usage Data' in the app settings.

9. Please provide a general description of how you will analyze this data.
	* By monitoring the count of how often users are performing these actions.

10. Where do you intend to share the results of your analysis?
	* With the mobile team.

11. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:
	* No.

**Comment by nbhasin2** [comment]
@boek Did some refactor and cleanup and removed code out of BVC

Also tagged Megan for Data review

**Comment by mmccorks** [comment]
1.    Is there or will there be documentation that describes the schema for the ultimate data set in a public, complete, and accurate way? 
Yes, this will be documented in docs/metrics.md

2.    Is there a control mechanism that allows the user to turn the data collection on and off? 
Yes, users can opt out of telemetry collection.

3.    If the request is for permanent data collection, is there someone who will monitor the data over time?
Not permanent data collection. 

4.    Using the category system of data types on the Mozilla wiki, what collection type of data do the requested measurements fall under?
Category 2, Interaction data.

5.    Is the data collection request for default-on or default-off?
6. Default on.

6.    Does the instrumentation include the addition of any new identifiers?
No new identifiers.

7.    Is the data collection covered by the existing Firefox privacy notice? 
Yes.

8.    Does there need to be a check-in in the future to determine whether to renew the data? 
Check in after 6 months (August 2021).

9.    Does the data collection use a third-party collection tool? 
No.


------
data-review +

**Review by mmccorks — COMMENTED** [review]
This looks good to me, once a search engineer has validated the telemetry implementation I am happy to sign off.

File: Client/Frontend/Home/FirefoxHomeViewController.swift [hunk 7]
```diff
@@ -581,11 +581,37 @@ extension FirefoxHomeViewController: DataObserverDelegate {
...
+            result.forEach {
+                if let _ = $0 as? PinnedSite {
+                    pinnedSites += 1
+                }
+            }
+            // Special case: Adding Google topsite
+            let googleTopSite = GoogleTopSiteHelper(prefs: self.profile.prefs)
+            let isIpad: Bool = UIDevice.current.userInterfaceIdiom == .pad
+            let isPortrait: Bool = UIApplication.shared.statusBarOrientation.isPortrait
+            let maxNumberOfTopSites = isIpad ? (isPortrait ? 12 : 16) : 8
```
Comment (q2r5) [inline Client/Frontend/Home/FirefoxHomeViewController.swift:597]: This breaks the setting for a custom number of top sites rows, `maxItems` should be used instead.

Comment (nbhasin2) [reply Client/Frontend/Home/FirefoxHomeViewController.swift:597]: Thanks for catching this @q2r5, going to push a chance to use `maxItems` that would avoid doing this logic entirely 🤞🏼

File: Client/Frontend/Home/GoogleTopSiteHelper.swift [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import Storage
+
+struct GoogleTopSiteConstants {
+    // A guid is required in the case the site might become a pinned site
+    public static let googleGUID = "DefaultGoogleGUID"
+    // US and rest of the world google urls
+    public static let usUrl = "https://www.google.com/webhp?client=firefox-b-1-m&channel=ts"
+    public static let rowUrl = "https://www.google.com/webhp?client=firefox-b-m&channel=ts"
+    public static let code = [ usUrl: "firefox-b-1-m",
+                               rowUrl: "firefox-b-m"]
```
Comment (Standard8) [inline Client/Frontend/Home/GoogleTopSiteHelper.swift:16]: I think this is unused.

Comment (nbhasin2) [reply Client/Frontend/Home/GoogleTopSiteHelper.swift]: > I think this is unused.

Yeah this one is not used. The one in Search Telemetry class is what we use

    private static let google = ["US": "firefox-b-1-m",
                                 "ROW": "firefox-b-m"]

**Pull request merged**